[VarDumper] Add support for options in `dump()`/`dd()`

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #48474 
| License       | MIT

Following up #48432, this PR brings an options builder to configure VarDumper's behavior thanks to a special `_options` named argument. Here is a snippet of it in action and a screenshot of the result. Feedbacks on UI are really welcomed, especially about the debug backtrace!

```php
<?php

use Symfony\Component\VarDumper\Dumper\CliDumper;

require 'vendor/autoload.php';

dump(
    var: ['Hi', 'Greetings' => 'Good morning, Symfony!'],
    _trace: true,
    _flags: CliDumper::DUMP_LIGHT_ARRAY | CliDumper::DUMP_STRING_LENGTH,
);
```

![image](https://github.com/user-attachments/assets/5633db23-e9a1-419a-88c5-b84f7e871f7d)

alexandre-daubois committed:
[VarDumper] Add support for options in `dump()`/`dd()`

nicolas-grekas reviewed (commented):
Thanks for tackling this, that's a complex feature because the matrix of possibilities is huge.

Here are some random thoughts:
- I like the style proposed in the issue: `dump()->trace()->maxDepth(4)->values($var);`. I'm just wondering if `values()` shouldn't be named `dump()` or `var()` or ?
- I also like the `dump($var, _trace: true)` style, which could work well with autocompletion if we describe the options as virtual arguments on the function? `array{_trace: bool, ...}`
- instead of trace_limit, what about `trace: bool|int`?
- have a look at `FrameStub`/`TraceStub` for dumping traces
- we might want to provide a way to define the default options on the VarDumper class?

lyrixx commented:
> * I like the style proposed in the issue: `dump()->trace()->maxDepth(4)->values($var);`. I'm just wondering if `values()` shouldn't be named `dump()` or `var()` or ?

[As states by @derrabus](https://github.com/symfony/symfony/issues/48474#issuecomment-1336581127), this cannot work

---

I don't really like the current object API:
```php
$options = (new VarDumperOptions())
    ->trace()
    ->stringLength()
;
```

it's too long to type. We should not forgot we are a in **debug** context. So we want to be able to get a result ASAP. that's why we have the function `dd()`. As short as possible.

That's why I prefer the second bullet point of @nicolas-grekas [comment](https://github.com/symfony/symfony/pull/48667#pullrequestreview-1220421921):


```php
dd($this, _trace: true, _maxDepth: 4);
```

lyrixx commented:
This is ugly :) I really don't like it. And it's too slow to type.

ro0NL commented:
how is a fully autodiscoverable `dump()->...` too slow to type?

lyrixx commented:
@ro0NL I mixed my thought :/ I was talking about: 

```php
$options = (new VarDumperOptions())
    ->trace()
    ->stringLength()
;

dump($this, _options: $options);
```

ro0NL commented:
> I'm just wondering if values() shouldn't be named dump() or var() or ?

maybe `__invoke` works, IIUC we could do

```
dump(1,2,3);
dump()(1,2,3)->trace();
dump()->trace()(1,2,3);
```

alexandre-daubois commented:
I agree the current object API is a bit much to type. I am :+1: to find a better solution.

Should I try to implement both @GromNaN solution (`dump` returning `mixed|VarDumperOptions`) and @nicolas-grekas' one, together? I wonder if it would be a bit disturbing to have both ways available, and I don't have, yet, a strong opinion for one solution over the other.

Also, I really like the solution proposed by @ro0NL. If I may find a drawback, I'm afraid this syntax could be difficult to be "discovered" by yourself without reading the documentation.

GromNaN commented:
The syntax that don't work is the one that rely on __destruct. 
The last suggestion mentioned by @nicolas-grekas is `dump()->trace()->maxDepth(4)->dump($var)`  which returns the `$var` as expected.

alexandre-daubois commented:
I updated the feature. The new version allows you to do both ways:

```php
// Passing options as named parameters
dump('Values', _trace: 5, _flags: AbstractDumper::DUMP_STRING_LENGTH);

// Use the options builder syntax
dump()
    ->trace(5)
    ->stringLength()
    ->dd('Values') // or ->dump('Values')
;
```

Still investigating on how to use `TraceStub`. Otherwise, this works well! And it is **so much cleaner** and quick to type.

Also, I made the choice to only have the options builder syntax on `dump`, not `dd`. I did this, because I think it is, in my opinion, a good thing to keep `never` return type on `dd`.

GromNaN commented on src/Symfony/Component/VarDumper/Resources/functions/dump.php:
```diff
@@ -9,28 +9,31 @@
+     *
+     * @return mixed|VarDumperOptions
      */
     function dump(mixed ...$vars): mixed
```
For IDE completion, it would be useful to add this return type.
```suggestion
    function dump(mixed ...$vars): mixed|VarDumperOptions
```

alexandre-daubois replied:
I wanted to, but PHPStorm told me that mixed type must be used as a standalone type and can't be used in an union type 😕

nicolas-grekas replied:
Does `@return mixed|VarDumperOptions` provide the target experience with phpstorm?

nicolas-grekas replied:
Could be `@template T` + `@param ...T` + `@return T|VarDumperOptions` work with both suggesting the options + the methods on the type passed as argument?

ro0NL replied:
https://psalm.dev/r/6b453ad4c5 :eyes:

alexandre-daubois replied:
Thank you both for helping with this, just learnt about template types!

@nicolas-grekas' solution works well, we get autocompletion of both the options and what's passed in argument. @ro0NL I tried to put your solution in PHPStorm, but it seems it doesn't understand this and autocompletion does not work, unfortunately 😕

ro0NL replied:
IIUC it's on their list https://youtrack.jetbrains.com/issue/WI-66465/Support-parsing-of-PhpStans-conditional-return-types

Perhaps keep a psalm-return annotations for reference, but ultimately whatever autocompletes in IDE works for me :)

ro0NL commented on src/Symfony/Component/VarDumper/VarDumper.php:
```diff
@@ -37,14 +42,25 @@ class VarDumper
+
     /**
      * @param string|null $label
+     * @param VarDumperOptions|iterable $options
```
s/iterable/array/

alexandre-daubois commented:
I've been able to use `TraceStub` to display the backtrace. Here is the result:

![image](https://user-images.githubusercontent.com/2144837/208432492-99d8d89a-f9c5-4067-a726-2b1176011517.png)

nicolas-grekas commented on src/Symfony/Bundle/DebugBundle/composer.json:
```diff
@@ -21,15 +21,16 @@
         "symfony/config": "<5.4",
-        "symfony/dependency-injection": "<5.4"
+        "symfony/dependency-injection": "<5.4",
+        "symfony/var-dumper": "<6.3"
```
not needed

nicolas-grekas commented on src/Symfony/Component/VarDumper/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * Add caster for `WeakMap`
  * Add support of named arguments to `dd()` and `dump()` to display the argument name
+ * Add support of dumper's behavior configuration with the special `_options` named argument for `dump` and `dd`
```
Let's keep thins simple, eg
Allowing configuring options when dumping

nicolas-grekas commented on src/Symfony/Component/VarDumper/Dumper/HtmlDumperTheme.php:
```diff
@@ -0,0 +1,21 @@
+/**
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ */
+enum HtmlDumperTheme: string
```
this should be removed as it closes the list of possible themes, while the list is open for extensibility

nicolas-grekas commented on src/Symfony/Component/VarDumper/Dumper/VarDumperOptions.php:
```diff
@@ -0,0 +1,193 @@
+        $this->options = array_replace($this->options, \array_intersect_key($options, $this->options));
+    }
+
+    public function dump(mixed ...$vars): mixed
```
would also benefit from the generics annotation

nicolas-grekas commented on src/Symfony/Component/VarDumper/Dumper/VarDumperOptions.php:
```diff
@@ -0,0 +1,193 @@
+            $vars = [new ScalarStub('🐛')];
+        }
+
+        return dump(...($vars + $this->options));
```
no need for the brackets

using dump() instead of VarDumper::dump() is adding one extra frame in the backtrace.
it might be better to skip this frame (same for dd() below

alexandre-daubois replied:
Did the update! Now, calling `dump` and `dd` through `VarDumperOptions` will now longer add a line to the backtrace.

nicolas-grekas commented on src/Symfony/Component/VarDumper/Dumper/VarDumperOptions.php:
```diff
@@ -0,0 +1,193 @@
+
+    public function dd(mixed ...$vars): never
+    {
+        dd(...($vars + $this->options));
```
no need for the brackets

nicolas-grekas commented on src/Symfony/Component/VarDumper/Dumper/VarDumperOptions.php:
```diff
@@ -0,0 +1,193 @@
+
+    public function __construct(array $options = [])
+    {
+        $this->options = array_replace($this->options, \array_intersect_key($options, $this->options));
```
why not just `$this->options = $options`?

alexandre-daubois replied:
In order to replace option values only by the given ones passed to the constructor. If an option is missing, it will still have a default value

alexandre-daubois replied:
Updated this a bit!

nicolas-grekas commented on src/Symfony/Component/VarDumper/Dumper/VarDumperOptions.php:
```diff
@@ -0,0 +1,193 @@
+    /**
+     * Display string lengths, just before its value.
+     */
+    public function stringLength(): static
```
I'd suggest prefixing all flag-related methods with "show": showStringLength, etc (or "display")

nicolas-grekas commented on src/Symfony/Component/VarDumper/Dumper/VarDumperOptions.php:
```diff
@@ -0,0 +1,193 @@
+        return $this->options[$option] ?? null;
+    }
+
+    public function getOptions(): array
```
toArray

nicolas-grekas commented on src/Symfony/Component/VarDumper/Dumper/VarDumperOptions.php:
```diff
@@ -0,0 +1,193 @@
+        return $this->options;
+    }
+
+    public function set(string $option, mixed $value): static
```
not needed I suppose

nicolas-grekas commented on src/Symfony/Component/VarDumper/Factory/VarClonerFactory.php:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @internal
+ */
+final class VarClonerFactory
```
I'd better not add this class, it adds indirection

nicolas-grekas commented on src/Symfony/Component/VarDumper/Dumper/VarDumperOptions.php:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    /**
+     * Set flags manually. Valid flags are {@see AbstractDumper::DUMP_*} constants.
```
`@param AbstractDumper::DUMP_* $flags`

nicolas-grekas commented on src/Symfony/Component/VarDumper/Resources/functions/dump.php:
```diff
@@ -9,28 +9,33 @@
-        if (!$vars) {
-            VarDumper::dump(new ScalarStub('🐛'));
+        $options = new VarDumperOptions($vars);
+        $vars = \array_diff_key($vars, $options->getOptions());
```
I'd rather ignore any arg that tarts with a `_` in the foreach below

nicolas-grekas commented on src/Symfony/Component/VarDumper/VarDumper.php:
```diff
@@ -37,14 +42,25 @@ class VarDumper
+        }
+
+        if (self::requiresRegister($options)) {
+            self::register($options);
```
this will change the options of the default handler when any options changes?
this would be unexpected to me.

alexandre-daubois replied:
`requiresRegister` helps on taking care of registering again the handler if `dump/dd` is not called with the exact same set of options as the previous call (this should work great with options defined in an env var!). So if you call the default dumper (I assume what you called "default dumper" is the one called without any special option? I may be wrong 😄), it should be registered again with its default options.

I struggled a bit on this part, only a few options need a full and new registration. I may be missing something?

nicolas-grekas reviewed (commented):
About the default options, I'd suggest configuring them via a VAR_DUMPER_OPTIONS env var. I'd go with a url-encoded string so that it's easy to turn into an options array

dunglas commented:
When dumping large objects, it could be nice to be able to have the possibility to dump only one or several properties instead of the full object. Something like: `_path: 'foo.bar'` and `_path: ['foo.bar', 'baz.bat']` maybe?

alexandre-daubois commented:
@nicolas-grekas thanky you for the review, I'll check this out! The env var is also a great idea too, in my opinion.

@dunglas I really like your proposition! Giving the size of this PR, I think it would deserve a separate one. Definitely keeping this in mind 👍

alexandre-daubois commented:
I updated the env var name to configure the behavior, from `VAR_DUMP_OPTIONS` to `VAR_DUMPER_OPTIONS`, to match with existing env var of the component (`VAR_DUMPER_FORMAT` & `VAR_DUMPER_SERVER`).

Also, I wrote the documentation about this PR, if you want to have a look at it: https://github.com/symfony/symfony-docs/pull/17787

alamirault commented on src/Symfony/Component/VarDumper/Dumper/VarDumperOptions.php:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    /**
+     * Display string lengths, just before its value.
```
```suggestion
     * Display string length, just before its value.
```

alexandre-daubois replied:
Thanks! :smile:

alamirault reviewed (commented):
Thanks Alexandre for this PR, I really like syntax with named parameters !

butschster commented:
Hey! We're currently working on [Buggregator](https://github.com/buggregator/server), and it can be used as a server for displaying dumps for the var-dumper component. I really like your PR with options. However, it would be great if you also provided the ability to add custom context as an array to a dump function.

```php
$options = (new VarDumperOptions())
    ->withContext(['language' => 'php'])
;

dump($this, _options: $options);
```

This feature can be used in various scenarios, such as providing information about syntax [highlighting](https://github.com/buggregator/server/issues/164) for string dumps or offering details about the project to which a dump should be linked.

Thanks for PR!

alexandre-daubois commented:
Hey @butschster! Thank you for your interest in this PR! I kinda forgot about it, but I still think it is worth it to have this in the framework.

About the context feature you're talking about: while I understand it would help in your case, I find it weird to add some context here that won't be used apart from super precise use cases. Two solutions here:

- If you have some strong reasons and use case, a complementary PR could be proposed ;
- We may keep `VarDumpOptions` **not** final so you can extend it for all your needs. I think that's the quick win 😊 

Anyway, I'll try finalize this PR. Thanks for digging it up!

nicolas-grekas commented on src/Symfony/Component/VarDumper/Tests/Dumper/functions/dump_without_args.phpt:
```diff
@@ -10,7 +10,7 @@ while (!file_exists($vendor.'/vendor')) {
 }
 require $vendor.'/vendor/autoload.php';
 
-dump();
```
by adding a destructor to the options class, we could (and should IMHO) keep dump() as is

nicolas-grekas commented on src/Symfony/Component/VarDumper/Dumper/VarDumperOptions.php:
```diff
@@ -0,0 +1,199 @@
+    {
+        $this->options = array_filter(
+            $options,
+            static fn (mixed $key): bool => \in_array($key, self::AVAILABLE_OPTIONS),
```
```suggestion
            static fn ($key) => \in_array($key, self::AVAILABLE_OPTIONS, true),
```

nicolas-grekas reviewed (commented):
Coming back after a few months here, I'm wondering if this is worth the extra complexity...
We have an unresolved concern, which is that dump-handlers don't have to react to options, and we don't have an internal API to communicate those options to them. If we do want options, we need that API.

I'm wondering if we need this options class. We could instead document options using `@param` on the dump() function, so that using them would look like: `dump($var, _max_depth: 3)`. That might be good enough in terms of DX.

alexandre-daubois commented:
I pushed a rebased, cleaner version. I removed the Options class and replaced it with an array, which seems enough. Tests are fixed.

> We have an unresolved concern, which is that dump-handlers don't have to react to options, and we don't have an internal API to communicate those options to them. If we do want options, we need that API.

I wasn't exactly sure what you meant here. I updated `VarDumper::dump()` in order to pass options to the registered dump handler. This way, if one defines its own handler, options can either be handled or ignored. Is that what you had in mind?

nicolas-grekas commented on src/Symfony/Component/VarDumper/Cloner/Data.php:
```diff
@@ -270,12 +271,17 @@ public function dump(DumperInterface $dumper): void
         $cursor->hashType = 0;
         $this->dumpItem($dumper, $cursor, $refs, $this->data[$this->position][$this->key]);
+
+        if (false !== ($options['_trace'] ?? false) && $cursor->attr = $this->context[BacktraceContextProvider::class] ?? []) {
```
comparisons to boolean... you missed testing this is `=== true` :P

alexandre-daubois replied:
Actually, `_trace` can be either a bool or a integer. `false` disables the trace. If it's an integer, we limit the trace size. Also, the limit is disabled if `_trace` is <= 0. That's why we can't use `=== true` here :smile: But maybe limiting the stack trace isn't that useful and we may only accept a boolean for `_trace`?

nicolas-grekas replied:
ah, makes sense then, let's keep as is :)

OskarStark commented:
PR title and header needs to be updated to reflect the content of this PR, thanks

alexandre-daubois commented:
Should be better now @OskarStark

nicolas-grekas commented:
Back on this, I think we should either close, or reduce to the strict minimum, which is that to me:

```
    /**
     * @author Nicolas Grekas <p@tchwork.com>
     *
     * @param int $_max_items
     * @param int $_min_depth
     * @param int $_max_string
     * @param int $_max_depth
     * @param int $_max_items_per_depth
     */
```

I'd add this docblock on the dump() and dd() functions to help discovery.
I'd remove the trace option: `dump(new \Exception())` does a better job already.
And the other options belong to global settings to me. We could have a mode-setting env var for them instead (but who's gonna change the defaults?)

nicolas-grekas commented:
@lyrixx can you play with and review #58070 then?

alexandre-daubois commented:
It's been 2 years and a half and I admit I didn't feel the need for options (neither met someone who needed this). I think `dd()` does already an amazing job and should be kept simple. That's why I'll close this PR. However if there's interest, I'll gladly continue this PR with Nicolas latest suggestions. Thanks for the time of everyone involved!
